lib/Infer/AbstractInterpreter.cpp

```diff
@@ -200,185 +217,236 @@ namespace souper {
+        Result.One.setBit(0);
+      if (getUMin(LHS).ugt(getUMax(RHS)))
+        Result.Zero.setBit(0);
       return Result;
     }
 
-    llvm::KnownBits sle(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+    llvm::KnownBits sle(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS) {
       llvm::KnownBits Result(1);
-      if (getSMax(lhs).sle(getSMin(rhs)))
-	Result.One.setBit(0);
-      if (getSMin(lhs).sgt(getSMax(rhs)))
-	Result.Zero.setBit(0);
+      if (getSMax(LHS).sle(getSMin(RHS)))
+        Result.One.setBit(0);
+      if (getSMin(LHS).sgt(getSMax(RHS)))
+        Result.Zero.setBit(0);
       return Result;
     }
   }
 
-  bool isReservedConst(Inst *I) {
-    return I->K == Inst::ReservedConst ||
-          (I->K == Inst::Var &&
-          (I->Name.find(ReservedConstPrefix) != std::string::npos));
-  }
-
-  bool isReservedInst(Inst *I) {
-    return I->K == Inst::ReservedInst ||
-          (I->K == Inst::Var &&
-          (I->Name.find(ReservedInstPrefix) != std::string::npos));
-  }
-
   bool isConcrete(Inst *I, bool ConsiderConsts, bool ConsiderHoles) {
     return !hasGivenInst(I, [ConsiderConsts, ConsiderHoles](Inst* instr) {
-			      if (ConsiderConsts && isReservedConst(instr))
-				return true;
-			      if (ConsiderHoles && isReservedInst(instr))
-				return true;
-
-			      return false;
-			    });
+      if (ConsiderConsts && isReservedConst(instr))
+        return true;
+      if (ConsiderHoles && isReservedInst(instr))
+        return true;
+      return false;
+    });
   }
 
-  EvalValue getValue(Inst *I, ValueCache &C, bool PartialEval) {
+  // Tries to get the concrete value from @I. If PartialEval is enabled, it tries to
+  // launch concrete interpreter to get the value. If not possible, it returns
+  // EvalValue of ValueKind::Unimplemented
+  EvalValue getValue(Inst *I, ConcreteInterpreter &CI, bool PartialEval) {
     if (I->K == Inst::Const)
       return {I->Val};
-    if (C.find(I) != C.end()) {
-      return C[I];
-    } else {
-      if (PartialEval && isConcrete(I)) {
-        return evaluateInst(I, C);
-      }
-    }
+    else if (I->K == Inst::Var && !isReservedConst(I))
+      // evaluateInst will only give us input value of the variable; it doesn't
+      // evaluate anything.
+      return CI.evaluateInst(I);
+
+    if (PartialEval && isConcrete(I))
+      return CI.evaluateInst(I);
+
     // unimplemented
     return EvalValue();
   }
 
-#define KB0 findKnownBits(I->Ops[0], C, PartialEval)
-#define KB1 findKnownBits(I->Ops[1], C, PartialEval)
-#define VAL(INST) getValue(INST, C, PartialEval)
+#define KB0 findKnownBits(I->Ops[0], CI, PartialEval)
+#define KB1 findKnownBits(I->Ops[1], CI, PartialEval)
+#define KB2 findKnownBits(I->Ops[2], CI, PartialEval)
+#define VAL(INST) getValue(INST, CI, PartialEval)
+
+  llvm::KnownBits mergeKnownBits(std::vector<llvm::KnownBits> Vec) {
+    assert(Vec.size() > 0);
+
+    auto width = Vec[0].getBitWidth();
```

Comment:
this is a pretty heavyweight kind of debugging code-- I'd say only do this when assertions are enabled or something

Reply:
makes sense. will modify.